### ISSUE TITLE
Duplicate symbols coming in `--emit-symbol-map`

### ISSUE THREAD

**Issue by ravisumit33** [emscripten-core/emscripten#19578]
Please include the following in your bug report:

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.35 (a90c5abb489c516f59c0d16864d953d4bedd37da)
clang version 17.0.0 (https://github.com/llvm/llvm-project 6865cff8ea8b07d9f2385fd92cecb422404f0f35)
Target: wasm32-unknown-emscripten
Thread model: posix
```

I am seeing duplicate symbols in the symbol map produced when using `--emit-symbol-map` flag. These symbols look like below:
```
std::exception::~exception()
std::exception::~exception().1
std::logic_error::~logic_error()
std::logic_error::~logic_error().1
```
There are many such symbols which ends with `.<number>` apart from the original  symbols. Are these really duplicate symbols? I am parsing symbol map file, can I ignore these and count them as original symbol?

**Comment by sbc100** [comment]
These are almost certainly real symbols that are generated by clang/llvm.   I can't remember why llvm does this but looking that the contents of the functions and the call sites will likely help figure out what they are.   Are you only seeing these extra symbols for destructors?   When you disassemble your wasm file (using `wasm-objdump -d`x or wasm-dis` for example) do the bodies of these functions look that same.

**Comment by sbc100** [comment]
So it looks like llvm generates 3 symbols with the same C++ name for `std::exception::~exception()`:

```
$ llvm-nm  cache/sysroot/lib/wasm32-emscripten/libc++abi.a  | grep logic_errorD
00000081 T _ZNSt11logic_errorD0Ev
00000001 T _ZNSt11logic_errorD1Ev
00000001 T _ZNSt11logic_errorD2Ev
../llvm-build/bin/llvm-nm  cache/sysroot/lib/wasm32-emscripten/libc++abi.a  | grep logic_errorD | c++filt 
00000081 T std::logic_error::~logic_error()
00000001 T std::logic_error::~logic_error()
00000001 T std::logic_error::~logic_error()
```

I suppose these are 3 different variants of the same destructor that clang uses internally in different situations.

They are different symbols with different mangled names but show up with the same demangled name, which is the name used by `--emit-symbol-map`.   I believe the `.1` is likely added by the binaryen's that dump the symbol map, in order to avoid giving more than one function that same name.

In other words, I think this is harmless/indented behaviour.

**Comment by ravisumit33** [comment]
Got it. Thanks.

**Comment by ravisumit33** [comment]
@sbc100 Can I get mangled functions with the `emit-symbol-map` flag so that I don't have to deal with these suffixes. It looks like there are a few kind of suffixes and writing regex for each of them doesn't seem to be a good idea.

**Comment by kripken** [comment]
Binaryen at least should use `.N` for such things. What other suffixes are you seeing @ravisumit33 ? Perhaps they arrive from LLVM.

**Comment by sbc100** [comment]
Is there some reason that you prefer the C++ mangling over the binaryen mangling? 

If so, we could look into adding some kind option to control this, but as with all new options, we would prefer to find a way to avoid adding them at all.

**Comment by ravisumit33** [comment]
@kripken I used these two regexes to find the actual symbol: `^(?P<symbol>.+?)(\.\d+)+$` & `^(?P<symbol>.*) \(\.\d+\)$`
There might be other patterns as well but in my codebase these two regexes worked properly.

**Comment by kripken** [comment]
@ravisumit33 Interesting, I have no idea where a suffix ` (N)` (space followed by a number in parentheses) could come from in that second pattern. @sbc100 does that ring a bell?

If not seeing some examples of such symbols might help. A testcase would be even better.

**Comment by ravisumit33** [comment]
@kripken Here is the symbol for second pattern : `std::__2::__fs::filesystem::detail::(anonymous namespace)::ErrorHandler<void>::report_impl(std::__2::error_code const&, char const*, void*) const (.368)`

Infact this is the only symbol that I encountered of this pattern.
